### DISCUSSION TITLE
JSON format documentation

### DISCUSSION THREAD

**Discussion by hydrargyrum** [tmo1/sms-ie#55] [General]
I'm the author of [Epistolaire](https://github.com/hydrargyrum/epistolaire/blob/master/backup.schema.json), an android app for exporting sms/mms to JSON. Unfortunately, it cannot import data yet.

To have better compatibility between our apps, I'm interested in what JSON format you use in your app. Epistolaire's format is described at https://github.com/hydrargyrum/epistolaire#json-format= and https://github.com/hydrargyrum/epistolaire/blob/master/backup.schema.json

Do you have a JSON schema description or a brief description? When importing, do you skip extra fields or do you reject the whole JSON?

This is related to another app which uses the same format as Epistolaire: https://github.com/hydrargyrum/breakthesilence/issues/1

**Comment by tmo1** [comment]
> I'm the author of [Epistolaire](https://github.com/hydrargyrum/epistolaire/blob/master/backup.schema.json), an android app for exporting sms/mms to JSON.

Welcome! Our apps are indeed similar, both in what they do, as well as in the philosophy behind them (data freedom and privacy, avoidance of the Play Store).

> Unfortunately, it cannot import data yet.

Had Epistolaire had import capability at the time I wrote SMS Import / Export, I might never have written it ;)

> Do you have a JSON schema description or a brief description?

No schema - description [here](https://github.com/tmo1/sms-ie#json-structure). One key difference between my JSON and yours is that I don't use conversation structures - perhaps I should. My thinking was that (IIUC) the notion of a "conversation" is not inherent to the messages themselves, but an organizational scheme layered on top of them by Android, but I can see that it might be helpful to preserve the Android conversation structure in the JSON.

I've been wanting to improve my JSON structure for a while, but I want to be quite sure before I take the plunge, since it's going to break compatibliity between versions (and / or require writing a migration tool), require a fair bit of careful recoding, and will likely introduce new bugs ;)

One thing I probably want to add is backup metadata (like your `errors` tag).

> When importing, do you skip extra fields or do you reject the whole JSON?

[The former](https://github.com/tmo1/sms-ie/commit/76b7b98f949ed82607b4f826907d6b327c3236dc).

> This is related to another app which uses the same format as Epistolaire: https://github.com/hydrargyrum/breakthesilence/issues/1

Ah, I see we also share qualms about Moxie's attitude and some of Signal's design decisions ;)

**Comment by hydrargyrum** [comment]
> Had Epistolaire had import capability at the time I wrote SMS Import / Export, I might never have written it ;)

Hehe, I'm a bit sorry I never took the time to implement it yet, but I'm happy to see other apps, more choice is better! And I should really give a try to SMS Import / Export app as it seems to fulfill several feature requests users made on Epistolaire.

> No schema - description here. One key difference between my JSON and yours is that I don't use conversation structures - perhaps I should. My thinking was that (IIUC) the notion of a "conversation" is not inherent to the messages themselves, but an organizational scheme layered on top of them by Android, but I can see that it might be helpful to preserve the Android conversation structure in the JSON.

I totally understand. My reasoning behind introducing conversations myself was that it would be easier to exploit the JSON for display for example, I hadn't importing in mind at that time (else, the decision would have probably been different).

> I've been wanting to improve my JSON structure for a while, but I want to be quite sure before I take the plunge, since it's going to break compatibliity between versions (and / or require writing a migration tool), require a fair bit of careful recoding, and will likely introduce new bugs ;)

Since SMS Import / Export app can import messages, I think I'll redirect some of my users to it (until I finally implement some import myself). I may write a conversion script for Epistolaire format to SMS Import / Export format, I did [a few converters](https://github.com/hydrargyrum/epistolaire/tree/master/converters) already.

> One thing I probably want to add is backup metadata (like your errors tag).

Dump timestamp etc.? About `errors`, it helps but not always, for example some users have encountered invalid JSON files with Epistolaire, and of course they can't share the data so it's very hard to reproduce :(

> > When importing, do you skip extra fields or do you reject the whole JSON?

> The former.

Great! That will avoid errors about futile conversion details.

> Ah, I see we also share qualms about Moxie's attitude and some of Signal's design decisions ;)

Now I'm laughing behind my computer, it seems we've had many common troubles :)

**Comment by tmo1** [comment]
> Great! That will avoid errors about futile conversion details.

My initial code just naively tried to insert all fields (except the ones we add on export). This worked fine in testing, since I was exporting and importing from AVDs running the same (or similar) versions of Android, but in the real world, people started seeing crashes (#11, #12, #14), so I realized that I had to do something about fields added in later Android versions. My solution was to just drop those fields and hope that the import still works - so far there have been no reports that it doesn't ;)

> Now I'm laughing behind my computer, it seems we've had many common troubles :)

I'm very unhappy about the phone number requirement, although they're apparently considering / working on changing that, and rather annoyed by the refusal to provide an unencrypted and documented backup / export of the message store, necessitating third party applications (which still only solves the encryption issue, not the lack of documentation).

I respect Moxie, both for his technical skills and for his good intentions, but I dislike his attitude of "we're going to do this the way we think is best, since we're the experts, regardless of user preferences" as well as the lack of interest or even hostility toward open standards and interoperability.

**Comment by gnusupport** [comment]
Please tell me definition of this time:
```
    "date": "1668776825112",
```

as it does not look as seconds from epoch.

**Reply by hydrargyrum** [reply]
It's the milliseconds since epoch.

**Reply by tmo1** [reply]
As @hydrargyrum says: see [here](https://developer.android.com/reference/android/os/SystemClock) and [here](https://developer.android.com/reference/java/sql/Date).

**Reply by gnusupport** [reply]
Thanks

**Comment by davebeep** [comment]
Hi. If at all possible, I am looking for help, please. I have a CSV file of old (10-15 years ago, pre-smart phone) SMS messages that I am trying to import to my current SMS app ([QKSMS](https://github.com/moezbhatti/qksms)). Each row is as follows (all punctuation as in original):
Row 1 of CSV-
**"type","address","body","date"**
For the avoidance of doubt, in my file **type** is 1 or 2 to show whether SMS was received or sent, **address** is a phone number beginning with either 0 or a +, **body** is the content of the SMS text, **date** is Unix time.
So, an example from row 2 downwards is-
2,"00447779877777","No, but I am leaving soon!","1120755000000"

How do I get this in to the correct format (such as, do the quotation marks need removing? or adding to the **type**?) and convert to a json that the app can read and export back out to my current SMS app? All my attempts have ended in failure. Or should I be using sms-db or something else?
Thanks very much!

**Reply by tmo1** [reply]
> Hi. If at all possible, I am looking for help, please. I have a CSV ...

I've opened [a separate issue](https://github.com/tmo1/sms-ie/issues/100) for this.
